### ISSUE TITLE
onchaind crashes with channel in ONCHAIN state; cannot recover funds

### ISSUE THREAD

**Issue by whitslack** [ElementsProject/lightning#4522]
### Issue and Steps to Reproduce
I had a locally funded channel with a peer I do not control. About two months ago, the peer initiated a mutual close of the channel. The channel state progressed from `CHANNELD_NORMAL` through `CHANNELD_SHUTTING_DOWN` to `CLOSINGD_SIGEXCHANGE`, where it remained until three days ago, when I attempted to force-close the channel. That transitioned the channel state through `AWAITING_UNILATERAL`, through `FUNDING_SPEND_SEEN`, to `ONCHAIN`, whereupon the onchain daemon crashed due to a segfault with instruction pointer at address zero. Looks like an indirect function call through a `NULL` function pointer.

The crash logged the following in the kernel log:
```
May 07 00:33:43 [kernel] lightning_oncha[27554]: segfault at 0 ip 0000000000000000 sp 00007ffe285b3008 error 14 in lightning_onchaind[562d0fc59000+7c000]
May 07 00:33:43 [kernel] Code: Unable to access opcode bytes at RIP 0xffffffffffffffd6.
```

I restarted C-Lightning a couple days later, and the `onchaind` process for that channel crashed again.
```
May 08 23:43:59 [kernel] lightning_oncha[12815]: segfault at 0 ip 0000000000000000 sp 00007fff111a7188 error 14 in lightning_onchaind[55ef5ff44000+7c000]
May 08 23:43:59 [kernel] Code: Unable to access opcode bytes at RIP 0xffffffffffffffd6.
```

Here's the (redacted) output from `listpeers`. Note the timestamps in the `state_changes` array:
```
{
   "peers": [
      {
         "id": "02…fe",
         "connected": false,
         "channels": [
            {
               "state": "ONCHAIN",
               "scratch_txid": "ac…29",
               "last_tx_fee": "1063000msat",
               "feerate": {
                  "perkw": 1467,
                  "perkb": 5868
               },
               "short_channel_id": "…13x…95x0",
               "direction": 0,
               "channel_id": "c1…4c",
               "funding_txid": "4c…c1",
               "close_to_addr": "bc1…jy",
               "close_to": "00…84",
               "private": false,
               "opener": "local",
               "closer": "remote",
               "features": [],
               "funding_allocation_msat": {
                  "02…fe": 0,
                  "02…2f": 1000000000
               },
               "funding_msat": {
                  "02…fe": "0msat",
                  "02…2f": "1000000000msat"
               },
               "msatoshi_to_us": 989829136,
               "to_us_msat": "989829136msat",
               "msatoshi_to_us_min": 20308403,
               "min_to_us_msat": "20308403msat",
               "msatoshi_to_us_max": 1000000000,
               "max_to_us_msat": "1000000000msat",
               "msatoshi_total": 1000000000,
               "total_msat": "1000000000msat",
               "fee_base_msat": "4815msat",
               "fee_proportional_millionths": 1,
               "dust_limit_satoshis": 546,
               "dust_limit_msat": "546000msat",
               "max_htlc_value_in_flight_msat": 18446744073709551615,
               "max_total_htlc_in_msat": "18446744073709551615msat",
               "their_channel_reserve_satoshis": 10000,
               "their_reserve_msat": "10000000msat",
               "our_channel_reserve_satoshis": 10000,
               "our_reserve_msat": "10000000msat",
               "spendable_msatoshi": 976696136,
               "spendable_msat": "976696136msat",
               "receivable_msatoshi": 170864,
               "receivable_msat": "170864msat",
               "htlc_minimum_msat": 0,
               "minimum_htlc_in_msat": "0msat",
               "their_to_self_delay": 144,
               "our_to_self_delay": 144,
               "max_accepted_htlcs": 483,
               "state_changes": [
                  {
                     "timestamp": "2021-03-03T10:50:21.472Z",
                     "old_state": "CHANNELD_NORMAL",
                     "new_state": "CHANNELD_SHUTTING_DOWN",
                     "cause": "remote",
                     "message": "Peer closes channel"
                  },
                  {
                     "timestamp": "2021-03-03T10:50:21.527Z",
                     "old_state": "CHANNELD_SHUTTING_DOWN",
                     "new_state": "CLOSINGD_SIGEXCHANGE",
                     "cause": "remote",
                     "message": "Start closingd"
                  },
                  {
                     "timestamp": "2021-05-07T04:23:49.490Z",
                     "old_state": "CLOSINGD_SIGEXCHANGE",
                     "new_state": "AWAITING_UNILATERAL",
                     "cause": "user",
                     "message": "Forcibly closed by `close` command timeout"
                  },
                  {
                     "timestamp": "2021-05-07T04:33:38.902Z",
                     "old_state": "AWAITING_UNILATERAL",
                     "new_state": "FUNDING_SPEND_SEEN",
                     "cause": "user",
                     "message": "Onchain funding spend"
                  },
                  {
                     "timestamp": "2021-05-07T04:33:43.645Z",
                     "old_state": "FUNDING_SPEND_SEEN",
                     "new_state": "ONCHAIN",
                     "cause": "user",
                     "message": "Onchain init reply"
                  },
                  {
                     "timestamp": "2021-05-09T03:43:48.430Z",
                     "old_state": "ONCHAIN",
                     "new_state": "FUNDING_SPEND_SEEN",
                     "cause": "user",
                     "message": "Onchain funding spend"
                  },
                  {
                     "timestamp": "2021-05-09T03:43:57.903Z",
                     "old_state": "FUNDING_SPEND_SEEN",
                     "new_state": "ONCHAIN",
                     "cause": "user",
                     "message": "Onchain init reply"
                  }
               ],
               "status": [
                  "ONCHAIN:Owning subdaemon onchaind died (-1)"
               ],
               "in_payments_offered": 350,
               "in_msatoshi_offered": 45154268462,
               "in_offered_msat": "45154268462msat",
               "in_payments_fulfilled": 144,
               "in_msatoshi_fulfilled": 14242962353,
               "in_fulfilled_msat": "14242962353msat",
               "out_payments_offered": 1332,
               "out_msatoshi_offered": 294393508686,
               "out_offered_msat": "294393508686msat",
               "out_payments_fulfilled": 37,
               "out_msatoshi_fulfilled": 14253133217,
               "out_fulfilled_msat": "14253133217msat",
               "htlcs": []
            }
         ]
      }
   ]
}
```

I have a core dump that yields the following backtrace in gdb:
```
Core was generated by `/usr/libexec/c-lightning/lightning_onchaind'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00007fd28f07ea75 in elf_nodebug (state=<optimized out>, pc=<optimized out>, callback=<optimized out>, 
    error_callback=<optimized out>, data=<optimized out>) at elf.c:533
#2  0x00007fd28f07e91e in unwind (context=<optimized out>, vdata=0x7fff111a7470) at backtrace.c:91
#3  0x00007fd28ec811f1 in _Unwind_Backtrace (trace=trace@entry=0x7fd28f07e8a0 <unwind>, 
    trace_argument=trace_argument@entry=0x7fff111a7470)
    at /var/tmp/portage/sys-devel/gcc-11.1.0/work/gcc-11.1.0/libgcc/unwind.inc:307
#4  0x00007fd28f07e993 in backtrace_full (state=0x7fd28f1f4000, skip=skip@entry=0, 
    callback=callback@entry=0x55ef5ff59820 <backtrace_status>, error_callback=error_callback@entry=0x0, data=data@entry=0x0)
    at backtrace.c:127
#5  0x000055ef5ff59c19 in send_backtrace (why=why@entry=0x7fff111a74b0 "FATAL SIGNAL 11") at common/daemon.c:44
#6  0x000055ef5ff59c67 in crashdump (sig=11) at common/daemon.c:52
#7  <signal handler called>
#8  0x0000000000000000 in ?? ()
#9  0x00007fd28f07ea75 in elf_nodebug (state=<optimized out>, pc=<optimized out>, callback=<optimized out>, 
    error_callback=<optimized out>, data=<optimized out>) at elf.c:533
#10 0x00007fd28f07e91e in unwind (context=<optimized out>, vdata=0x7fff111a7c90) at backtrace.c:91
#11 0x00007fd28ec811f1 in _Unwind_Backtrace (trace=trace@entry=0x7fd28f07e8a0 <unwind>, 
    trace_argument=trace_argument@entry=0x7fff111a7c90)
    at /var/tmp/portage/sys-devel/gcc-11.1.0/work/gcc-11.1.0/libgcc/unwind.inc:307
#12 0x00007fd28f07e993 in backtrace_full (state=0x7fd28f1f4000, skip=skip@entry=0, 
    callback=callback@entry=0x55ef5ff59820 <backtrace_status>, error_callback=error_callback@entry=0x0, data=data@entry=0x0)
    at backtrace.c:127
#13 0x000055ef5ff59c19 in send_backtrace (why=why@entry=0x55ef61448d78 "Could not find resolution for output 0")
    at common/daemon.c:44
#14 0x000055ef5ff624eb in status_failed (reason=reason@entry=STATUS_FAIL_INTERNAL_ERROR, 
    fmt=fmt@entry=0x55ef5ffa04e0 "Could not find resolution for output %zu") at common/status.c:207
#15 0x000055ef5ff54252 in handle_our_unilateral (tx=tx@entry=0x55ef6136ad08, tx_blockheight=682337, 
    basepoints=basepoints@entry=0x7fff111a8210, htlcs=htlcs@entry=0x55ef6136c498, 
    tell_if_missing=tell_if_missing@entry=0x55ef6137cb98, tell_immediately=tell_immediately@entry=0x55ef6137d268, 
    remote_htlc_sigs=0x0, outs=<optimized out>, is_replay=true) at onchaind/onchaind.c:2798
#16 0x000055ef5ff4b927 in main (argc=<optimized out>, argv=<optimized out>) at onchaind/onchaind.c:3853
```

### `getinfo` output
```
{
   "id": "…",
   "alias": "…",
   "color": "…",
   "num_peers": …,
   "num_pending_channels": 2,
   "num_active_channels": …,
   "num_inactive_channels": 8,
   "address": [
      {
         "type": "ipv4",
         "address": "…",
         "port": 9735
      },
      {
         "type": "ipv6",
         "address": "…",
         "port": 9735
      }
   ],
   "binding": [
      {
         "type": "ipv6",
         "address": "::",
         "port": 9735
      },
      {
         "type": "ipv4",
         "address": "0.0.0.0",
         "port": 9735
      }
   ],
   "version": "0.10.0",
   "blockheight": 682965,
   "network": "bitcoin",
   "msatoshi_fees_collected": …,
   "fees_collected_msat": "…msat",
   "lightning-dir": "/var/lib/lightning/bitcoin"
}
```

**Comment by niftynei** [comment]
Just wanted to point out that this is the key line from the stack trace, that explains why the node fell over:

```
#13 0x000055ef5ff59c19 in send_backtrace (why=why@entry=0x55ef61448d78 "Could not find resolution for output 0")
    at common/daemon.c:44
```

**Comment by whitslack** [comment]
> Was this node restored from a backup?

No, not restored from a backup. It *is* a rather old node, though. The particular channel in question was opened in August 2019.

> We should never get into a situation in which we cannot resolve an output of the channel, unless our DB was somehow manipulated and we lost past states.

What is the meaning of "resolve the output of the channel"? Does that mean my node wants to find a transaction that it can publish to "claim" the "to self" output of the channel, but it doesn't know how to do that?

> The crash is on purpose to make as much noise as possible, and get an operator to notice it btw :-)

I certainly noticed, though I'm not sure that segfaulting is a great way to abort a program. It's like shutting off the engine in your car by throwing a wrench into the belts and pulleys. It'll make a hell of a noise, but it's not exactly good for the engine.

**Upon further examination,** it appears that the "to self" output from this channel's commitment transaction *was* actually claimed by my node, and the UTxO resulting from that claim was subsequently used in a funding transaction to open some more channels. So apparently this channel is just hanging around needlessly now. Can I make it disappear?

**Comment by whitslack** [comment]
Maybe this is a relevant observation, and I think there may really be a bug here…

The transaction that was actually published to the blockchain to close this channel appears to be a **mutual**-close transaction, as both outputs are P2WPKH. I was mistaken in my previous comment: there actually was no claim of the output of the commitment transaction because it was actually not a commitment transaction that closed the channel but rather a mutual close. The output for my side of the channel went straight into my node's wallet to be used by a future channel funding transaction.

So, I think what happened is this:

1. On 3 March, the peer initiated a mutual close.
2. Our nodes successfully constructed and signed a closing transaction, but it was not published to the blockchain for some unknown reason.
3. The peer disappeared.
4. On 7 May, I attempted to *unilaterally* close the limbo'd channel.
5. Amazingly, my node must have published the previously negotiated *mutual* closing transaction, **not** the commitment transaction.
6. The output from the mutual closing transaction went straight into my node's wallet, but because I had issued a unilateral close command, my node started trying to claim a non-existent time-locked output from a commitment transaction that never hit the blockchain.
